dashboard/modules/job/job_agent.py

```diff
@@ -0,0 +1,303 @@
+        await self._unpack_package(filename, unpack_dir)
+
+
+class StartPythonDriver(JobProcessor):
+    _template = """import sys
+sys.path.append({import_path})
+import ray
+from ray._private.utils import hex_to_binary
+ray.init(ignore_reinit_error=True,
+         address={redis_address},
+         _redis_password={redis_password},
+         job_id=ray.JobID({job_id}),
+         job_config=ray.job_config.JobConfig({job_config_args}),
+)
+import {driver_entry}
+{driver_entry}.main({driver_args})
+# If the driver exits normally, we invoke Ray.shutdown() again
+# here, in case the user code forgot to invoke it.
+ray.shutdown()
+"""
+
+    def __init__(self, job_info, redis_address, redis_password):
+        super().__init__(job_info)
+        self._redis_address = redis_address
+        self._redis_password = redis_password
+
+    def _gen_driver_code(self):
+        temp_dir = self._job_info.temp_dir
+        job_id = self._job_info.job_id
+        job_package_dir = job_consts.JOB_UNPACK_DIR.format(
+            temp_dir=temp_dir, job_id=job_id)
+        driver_entry_file = job_consts.JOB_DRIVER_ENTRY_FILE.format(
+            temp_dir=temp_dir, job_id=job_id, uuid=uuid.uuid4())
+        ip, port = self._redis_address
+
+        # Per job config
+        job_config_items = {
+            "worker_env": self._job_info.env,
+            "code_search_path": [job_package_dir],
+        }
+
+        job_config_args = ", ".join(f"{key}={repr(value)}"
+                                    for key, value in job_config_items.items()
+                                    if value is not None)
+        driver_args = ", ".join([repr(x) for x in self._job_info.driverArgs])
+        driver_code = self._template.format(
+            job_id=repr(hex_to_binary(job_id)),
+            job_config_args=job_config_args,
+            import_path=repr(job_package_dir),
+            redis_address=repr(ip + ":" + str(port)),
+            redis_password=repr(self._redis_password),
+            driver_entry=self._job_info.driverEntry,
+            driver_args=driver_args)
+        with open(driver_entry_file, "w") as fp:
+            fp.write(driver_code)
+        return driver_entry_file
+
+    async def run(self):
+        python = self._get_current_python()
+        driver_file = self._gen_driver_code()
+        driver_cmd = [python, "-u", driver_file]
+        stdout_file, stderr_file = self._new_log_files(
+            self._job_info.log_dir, f"driver-{self._job_info.job_id}")
+        return await self._start_driver(driver_cmd, stdout_file, stderr_file,
+                                        self._job_info.env)
+
+
+class JobAgent(dashboard_utils.DashboardAgentModule,
+               job_agent_pb2_grpc.JobAgentServiceServicer):
+    def __init__(self, dashboard_agent):
+        super().__init__(dashboard_agent)
+        self._job_table = {}
+
+    async def InitializeJobEnv(self, request, context):
```

Comment:
this should be snake_case, right?

Reply:
This is the RPC API, it is defined in the protobuf. We use the Camel-Case to make all RPC styles consistent.